test/adapter/json_api/relationship_test.rb

```diff
@@ -50,126 +46,350 @@ def test_relationship_with_data_array
               }
             ]
           }
-          test_relationship(expected, options: { include_data: true })
+
+          model_attributes = { posts: [Post.new(id: 1), Post.new(id: 2)] }
+          relationship_name = :posts
+          model = new_model(model_attributes)
+          actual = build_serializer_and_serialize_relationship(model, relationship_name) do
+            has_many :posts
+          end
+          assert_equal(expected, actual)
         end
 
         def test_relationship_data_not_included
-          test_relationship({ meta: {} }, options: { include_data: false })
-        end
+          expected = { meta: {} }
 
-        def test_relationship_simple_link
-          links = { self: 'a link' }
-          test_relationship({ links: { self: 'a link' } }, links: links)
+          model_attributes = { blog: :does_not_matter }
+          relationship_name = :blog
+          model = new_model(model_attributes)
+          actual = build_serializer_and_serialize_relationship(model, relationship_name) do
+            has_one :blog do
+              include_data false
+            end
+          end
+          assert_equal(expected, actual)
         end
 
         def test_relationship_many_links
-          links = {
-            self: 'a link',
-            related: 'another link'
-          }
           expected = {
             links: {
               self: 'a link',
               related: 'another link'
             }
           }
-          test_relationship(expected, links: links)
-        end
 
-        def test_relationship_block_link
-          links = { self: proc { object.id.to_s } }
-          expected = { links: { self: @blog.id.to_s } }
-          test_relationship(expected, links: links)
+          model_attributes = { blog: :does_not_matter }
+          relationship_name = :blog
+          model = new_model(model_attributes)
+          actual = build_serializer_and_serialize_relationship(model, relationship_name) do
+            has_one :blog do
+              include_data false
+              link :self, 'a link'
+              link :related, 'another link'
+            end
+          end
+          assert_equal(expected, actual)
         end
 
         def test_relationship_block_link_with_meta
-          links = {
-            self: proc do
-              href object.id.to_s
-              meta(id: object.id)
-            end
-          }
           expected = {
             links: {
               self: {
-                href: @blog.id.to_s,
-                meta: { id: @blog.id }
+                href: '1',
+                meta: { id: 1 }
               }
             }
           }
-          test_relationship(expected, links: links)
+
+          model_attributes = { blog: Blog.new(id: 1) }
+          relationship_name = :blog
+          model = new_model(model_attributes)
+          actual = build_serializer_and_serialize_relationship(model, relationship_name) do
```

Comment:
I really like this pattern, of defining the serializer for test in the test itself